Refactor Supply::Client to be usable by create_app_on_managed_play_store

janpio opened the pull request:
`create_app_on_managed_play_store` does very similar things as `supply`, so effectively had a copy of the API communication logic that was worse than what `supply` had.

This PR extracts parts of `Supply::Client` into `Supply::GoogleServiceClient` which `Supply::Client` then inherits from and which can be used by `create_app_on_managed_play_store` with its own implementation.

--- 

Some notes:

- I kept the handling of deprecated params in Supply::Client only
- The refactor decouples `make_from_config` and `initialize` from `Supply.config` a bit so it doesn't _have to_ be used



There are some icky areas:

- ~~The idea was, that `scope` and `service` could be overwritten or set in the child class for other uses of the client. But `@scope` and `@service` didn't work if I just set them in the class, I had to overwrite their values in `initialize`.~~ Fixed.

- During development I noticed that `supply` is almost untested. I could do pretty much what I wanted in the client, the tests still passed. But I manually tested this branch now (`bundle exec fastlane suppy init`, lane that uses `supply` to upload downloaded metadata to another app), and it seems to work fine)

janpio committed:
fix order in verify_block for Supply's json_key:
check if file exists first, then if it is JSON

janpio committed:
Extract reusable parts of Client into GoogleServiceClient (self.make_from_config, initialize, call_google_api)

janpio committed:
Modify create_app_on_managed_play_store to implement and use Supply::GoogleServiceClient

janpio committed:
remove leftover code and TODOs
rubocop

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Pull Request is Work in Progre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

janpio committed:
skip fastlane/spec/unused_options_spec.rb for new action

janpio committed:
remove TODO

joshdholtz commented on supply/lib/supply/client.rb:
```diff
@@ -5,7 +5,67 @@
+
+    # Initializes the service and its auth_client using the specified information
+    # @param service_account_json: The raw service account Json data
+    def initialize(service_account_json: nil, params: nil)
```
(see comment below this first)

It might also be nice to add `scope` and `service` as a parameter to this `initialize` method so that subclasses know that they need to be passed instead/set.

janpio replied:
I had problems with subclasses having different parameters on methods - but that was probably me not knowing how ruby works. (I guess I messed up the method param signatures or something, `params` vs. `params:nil` etc)

joshdholtz reviewed (dismissed):
Added some comments about implementation of `GoogleServiceClient`. Let me know if you have any questions on my questions 🙃

janpio commented on supply/lib/supply/client.rb:
```diff
@@ -5,7 +5,67 @@
+    # Initializes the service and its auth_client using the specified information
+    # @param service_account_json: The raw service account Json data
+    def initialize(service_account_json: nil, params: nil)
+      scope = @scope
```
`AbstractGoogleServiceClient` sounds good - the base class should not be usable by itself as it is missing the scope and service config.

`@scope` and `@service` being set in `initialize` is an ugly workaround.  

What I actually wanted to have is a base class and two implementations for it, that are "configured" differently. Just using one should have the correct service and scope, using the other would have its own scope and service set.

I tried to set `@scope = 'foo'` on the class level but that didn't work, so I (temp) moved it into `initialize`, which leads to all those problems.

How can I achieve this?

joshdholtz replied:
I think maybe you want something like this? https://stackoverflow.com/a/19382692/2464643

```rb
class Foo
  TEST = "a"

  def speak
    puts self.class::TEST
  end
end

class Bar < Foo
  TEST = "b"
end

Bar.new.speak # => a
Foo.new.speak # => b
```

janpio replied:
I saw that, but it looked to strange at the moment.
But yep, that could work. What value would those have in my abstract class? Be empty? Default of the Supply::Client there as well?

joshdholtz replied:
They would probably be `nil` in your abstract class and the you would probably need to just add an extra check/verification in your abstract class to make sure they are set because otherwise 💥 

This is a bit weird which is why the approach I would probably go with is passing them into the initialize parameter as this feel like a dependency injection thing. Passing them into the initialize parameter makes the most sense to me since the abstract class doesn't have any of its own so it feels weird to define a class level variable for them. 🙃

janpio replied:
So they would be defined and stored in `make_from_config` and then given to `new` as a param? (That feels pretty wrong as well)

janpio replied:
At least it will be obvious and pretty clear.

janpio replied:
Both `AbstractGoogleServiceClient` and `self.class::SERVICE`/`self.class::SCOPE` are now used.

janpio committed:
Rename Supply::GoogleServiceClient to Supply::AbstractGoogleServiceClient

janpio committed:
replce @scope/@service with SCOPE/SERVICE and use self.class::SCOPE/SERVICE to read it
https://stackoverflow.com/questions/19382626/how-do-i-override-a-variable-in-a-ruby-subclass-without-affecting-the-superclass/19382692#19382692

janpio committed:
workaround for `warning: already initialized constant Supply::UI`

janpio committed:
get rid of more code by moving client/android_publisher to AbstractGoogleServiceClient as client
(and also get rid of empty method implementations in PlaycustomappClient)

janpio commented:
This is read for another look. Implemented the comments from @joshdholtz and improved the code a bit more now that the class hierarchy makes more sense.

joshdholtz commented on supply/lib/supply/client.rb:
```diff
@@ -5,9 +5,74 @@
+  end
+
+  class Client < AbstractGoogleServiceClient
+    SERVICE = Androidpublisher::AndroidPublisherService.new
```
I don't know if we want to actually call `.new`here. There is a line in `AbstractGoogleServiceClient` that does the following...

```rb
service = self.class::SERVICE
service.authorization = auth_client
```

This will modify the class `SERVICE` class variable. I THINK what we want to do instead is change `SERVICE` to just be `SERVICE = Androidpublisher::AndroidPublisherService` and then create a new method in `AbstractGoogleServiceClient` that creates the instance like...

```rb
def service
  @service ||= self.class::self.class::SERVICE
  return @service
end
```

Or something like this? Let me know if my concern sounds crazy 🙃

janpio replied:
Oh right, that first line doesn't create a copy.

But you probably meant:
```
  @service ||= self.class::SERVICE.new
```
Correct?

We could also just do `service = self.class::SERVICE.new` possibly and not add `@service` into the mix?

joshdholtz replied:
Depends 😊 `@service ||= self.class::SERVICE.new` has a benefit of the service only creating one single instance so that you can call the `service` method multiple times without needing to worry about accidentally creating multiple instances

joshdholtz replied:
Ehhh, ish... `attr_accessor` automatically creates the `@variable_name` and the getter and setter methods that go along with it. This is a little bit different because this method is just a getter that is initializing the `@service` if it hasn't been initialized and then returning it

joshdholtz reviewed (dismissed):
Left some comments about making `SERVICE` just a reference to a class and then creating the instance from that (to prevent modification on a class variable)

janpio committed:
move .new into initialize()

joshdholtz reviewed (approved):
👍
